nvflare/job_config/stats_job.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List
+
+from nvflare import FedJob, FilterType
+from nvflare.app_common.abstract.statistics_spec import Statistics
+from nvflare.app_common.executors.statistics.statistics_executor import StatisticsExecutor
+from nvflare.app_common.filters.statistics_privacy_filter import StatisticsPrivacyFilter
+from nvflare.app_common.statistics.histogram_bins_cleanser import HistogramBinsCleanser
+from nvflare.app_common.statistics.json_stats_file_persistor import JsonStatsFileWriter
+from nvflare.app_common.statistics.min_count_cleanser import MinCountCleanser
+from nvflare.app_common.statistics.min_max_cleanser import AddNoiseToMinMax
+from nvflare.app_common.workflows.statistics_controller import StatisticsController
+
+
+class StatsJob(FedJob):
+    def __init__(
+        self,
+        job_name: str,
+        statistic_configs: dict,
+        stats_generator: Statistics,
+        output_path: str,
+        min_count: int = 10,
+        min_noise_level=0.1,
+        max_noise_level=0.3,
+        max_bins_percent=10,
+    ):
+        super().__init__()
+        self.writer_id = "stats_writer"
+        self.stats_generator_id = "stats_generator"
+        self.job_name = job_name
+        self.stats_generator = stats_generator
+        self.statistic_configs = statistic_configs
+        self.output_path = output_path
+        self.min_count = min_count
+        self.min_noise_level = min_noise_level
+        self.max_noise_level = max_noise_level
+        self.max_bins_percent = max_bins_percent
+
+        self.setup_server()
+
+    def setup_server(self):
+        # define stats controller
+        ctr = self.get_stats_controller()
+        # define stat writer to output Json file
+        stats_writer = self.get_stats_output_writer()
+        self.to(ctr, "server")
+        self.to(stats_writer, "server", id=self.writer_id)
+
+    def setup_client(self, sites: List[str]):
+        # Client side job config
+        # Add client site
+        for site_id in sites:
+            self.to(self.stats_generator, site_id, id=self.stats_generator_id)
+            executor = StatisticsExecutor(generator_id=self.stats_generator_id)
```

Comment:
The "to" can actually use different id other than the specified one.

It should be

```
real_id = self.to(self.stats_generator, site_id, id=self.stats_generator_id)
executor = StatisticsExecutor(generator_id=real_id)
```

Reply:
yes, this can occur if multiple components with id self.stat_generator_id ("stats_generator") are added, as the subsequent ones will use incremented ids "stat_generator1/2/3"